Add tests for directv platform

ehendrix23 opened the pull request:
## Description:

Create test for DirecTV platform.
Added media_stop to common for media player

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

ehendrix23 committed:
Create test for platform

Created test for platform.
Added media_stop to common.py test

ehendrix23 committed:
Multiple improvements

Fixed lint issue in common.py
Fixed lint issues in test_directv.py
Improved patching import using modile_patcher.start() and stop()
Added asserts for service calls.

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,445 @@
+        self.attributes['major'] = int(source)
+
+
+class TestSetupDirectvMediaPlayer(unittest.TestCase):
```
We prefer standalone pytest test functions for new tests. See eg the deconz test suite.

ehendrix23 replied:
@MartinHjelmare , all changed. Let me know. :-)

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,445 @@
+        modules = {
+            'DirectPy': self.directpy_mock
+        }
+        self.module_patcher = patch.dict('sys.modules', modules)
```
Use MockDependency in tests/common.py.

ehendrix23 replied:
Done. :-)

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,445 @@
+    def test_setup_platform_config(self):
+        """Test setting up the platform from configuration."""
+        add_entities = Mock()
+        setup_platform(
```
Don't set up the platform directly. Set up the component with `async_setup_component` and pass the config with the media_player component and the directtv platform.

ehendrix23 replied:
Done :-)

ehendrix23 committed:
Updates based on Martin's review

Updates based on Martin's review.

ehendrix23 commented:
@MartinHjelmare ,

All changed accordingly, let me know how it looks. This is the 1st test I'm writing. :-)

ehendrix23 committed:
Updated test based on PR#18474

Updated test to use service play_media instead of select_source based on change from PR18474

ehendrix23 committed:
Lint issues

Lint issues

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,494 @@
+    "rating": "No Rating",
+    "startTime": 1541876400,
+    "stationId": 3900947,
+    "title": "CNN Newsroom With Fredricka Whitfield"
```
Can we use a fake name here?

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,494 @@
+        self._standby = True
+        self._play = False
+
+        _LOGGER.debug("INIT with host: %s, port: %s, device: %s",
```
Don't log in tests.

ehendrix23 replied:
Yeah, was doing that as I was testing the test. :-)

Will remove.

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,494 @@
+    with MockDependency('DirectPy'), \
+            patch('DirectPy.DIRECTV', new=mockDIRECTVClass):
+
+        setup_platform(hass, {}, add_entities,
```
Set up the component but without the platform config and then fire the correct discovery event to test discovery.

MartinHjelmare replied:
We might need to set up discovery component and not the media_player component.

MartinHjelmare replied:
Nah, let's just call `async_load_platform` from helpers.discovery.py.

MartinHjelmare replied:
An empty dict as fifth argument to `async_load_platform` should work as hass config in this case.

ehendrix23 replied:
Just seeing your message. Not passing an empty dict but instead just one with component in it.
Can change to empty if you want me to. Let me know. :-)

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,494 @@
+        }]
+    }
+
+    async def setup_test_platforms(hass):
```
Why make a factory? Instead `yield` from the fixture to keep the patch in place during the test.

ehendrix23 replied:
I used it to be able to return the entity objects. Changed now.

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,494 @@
+    main_media_entity, client_media_entity = await platforms(hass)
+
+    # Features supported for main DVR
+    assert mp.SUPPORT_PAUSE | mp.SUPPORT_TURN_ON | mp.SUPPORT_TURN_OFF |\
```
Assert state attributes instead of entity attributes.

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,494 @@
+        # Turn main DVR on. When turning on DVR is playing.
+        await async_turn_on(hass, MAIN_ENTITY_ID)
+        await hass.async_block_till_done()
+        mock_key_press.assert_called_with('poweron')
```
Don't use mock assert methods. A simple typo in the method name could make the assertion pass when it should fail. Import mock.call helper method and assert the `call_args` attribute on the mock instance.

MartinHjelmare replied:
If we do a typo in the mock assert method name, it will be interpreted as a call to a mock mock method which will always pass. That can't happen in the latter case.

ehendrix23 replied:
Thus you mean if we do a typo within the test itself. For example:
`mock_key_press.assert_caled_with('poweron')`
Then the assert will pass no matter what. 

I just tested this.
`mock_key_press.assert_caled_with('poweron')`
fails with the error:
`AttributeError: assert_caled_with`

And if I try:
`mock_key_pres.assert_called_with('poweron')`
then it fails with `NameError: name 'mock_key_pres' is not defined`

And type in the patch.object itself (using 'key_pres' instead of 'key_press' gives me:
`AttributeError: <tests.components.media_player.test_directv.MockDirectvClass object at 0x106ddd898> does not have the attribute 'key_pres'`

And if I try:
`mock_key_press.asert_called_with('poweron')`
then I get: AttributeError: 'function' object has no attribute 'asert_called_with'

Not sure if something has changed. I did find:
https://engineeringblog.yelp.com/2015/02/assert_called_once-threat-or-menace.html

But I tried what they did there and it failed (assert_called_once in their example does work but it is also a valid assert from Python 3.6 onwards).

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,494 @@
+    main_media_entity, client_media_entity = await platforms(hass)
+
+    # Confirm service is currently set to available.
+    assert main_media_entity.available
```
Assert state attributes, not entity attributes.

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,494 @@
+        client_media_entity.schedule_update_ha_state(True)
+        await hass.async_block_till_done()
+
+        return (main_media_entity, client_media_entity)
```
Don't return the entities. We should use the states from the state machine to assert state in the tests.

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,494 @@
+    await hass.services.async_call(DOMAIN, SERVICE_PLAY_MEDIA, data)
+
+
+class mockDIRECTVClass:
```
`MockDIRECTV`

https://www.python.org/dev/peps/pep-0008/#class-names

ehendrix23 replied:
Based on that, should be: MockDirectvClass?

FYI, reason why I called it mockDIRECTVClass is cause the name of the original class is DIRECTV :-)

ehendrix23 committed:
Further updates based on feedback

Updates based on feedback provided.

ehendrix23 committed:
Using async_load_platform for discovery test

Using async_load_platform for discovery tests.
Added asserts to ensure entities are created with correct names.

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,507 @@
+        await async_setup_component(hass, mp.DOMAIN, WORKING_CONFIG)
+        await hass.async_block_till_done()
+
+    assert sum(1 for item in hass.data['media_player'].entities) == 1
```
Don't assert things about entitites. Assert states for the corresponding entity_id or domain.

ehendrix23 replied:
What would be the best method then to check the total number of entities within a component? Idea here is to make sure that only 1 entity was created.

MartinHjelmare replied:
Get all states for the domain.

ehendrix23 replied:
Well, that did not mean I knew how to do it. :-)

Found the StateMachine class in the core, able to figure it out from there. :-)

One thing for sure, also learning a lot more about the inners of HASS and how things are distributed throughout the files <g>.

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,507 @@
+
+
+@pytest.fixture
+async def platforms(hass):
```
Standard pytest fixtures can't be coroutines. If we need to schedule a coroutine in a fixture, use `loop.run_until_complete`.

ehendrix23 replied:
If I use:
```
setup_component(hass, mp.DOMAIN, config)
hass.block_till_done()
```
to setup the component within the fixture then it hangs.

If I use:
```
await async_setup_component(hass, mp.DOMAIN, config)
await hass.async_block_till_done()
```
then it works but only if I use:
`async def platforms(hass)`

What is the right method then to setup a component within a fixture?

MartinHjelmare replied:
```py
hass.loop.run_until_complete(
    async_setup_component(...))
```

ehendrix23 replied:
Think I figured it out. For some reason I always seem to get to the answer right after I ask here. Next time just going to ask before I start looking and wasting time, then once asked the answer will appear on my screen right away. :-)

MartinHjelmare commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,507 @@
+    client_media_entity.schedule_update_ha_state(True)
+    await hass.async_block_till_done()
+
+    return platforms
```
Don't return the fixture.

ehendrix23 committed:
Used HASS event_loop to setup component

Use HASS event_loop to setup the component async.

ehendrix23 committed:
Updated to use state machine for # entities

Updated to use state machine to count # entities instead of entities.

ehendrix23 commented:
@MartinHjelmare OK, did those updates as well. :-) Been quite a learning experience. <g>.

Review and let me know if there is anything else. :-)

ehendrix23 committed:
Use hass.loop instead of getting current loop

Small update to use hass.loop instead, thanks Martin!

ehendrix23 committed:
Forgot to remove asyncio

Removed asyncio import.

ehendrix23 committed:
Added fixtures

Added fixtures.

ehendrix23 commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,501 @@
+        await hass.async_block_till_done()
+
+    assert len(hass.states.async_entity_ids('media_player')) == 1
+    assert hass.data['media_player'].get_entity(MAIN_ENTITY_ID) is not None
```
Changed to get the state for the entity and then assert if state exist. If it does then we know the entity was indeed created with the correct name.

ehendrix23 commented on tests/components/media_player/test_directv.py:
```diff
@@ -0,0 +1,501 @@
+        CLIENT_ENTITY_ID)
+
+    # Set the client so it seems a recording is being watched.
+    client_media_entity.dtv.attributes = RECORDING
```
Made the changed and been going over it for a while to understand. However this is not working right now.
I get the following error:
```
ERROR:homeassistant.helpers.entity:Update for media_player.main_dvr fails
Traceback (most recent call last):
  File "/Users/ehendrix/Documents/GitHub/home-assistant/homeassistant/helpers/entity.py", line 221, in async_update_ha_state
    await self.async_device_update()
  File "/Users/ehendrix/Documents/GitHub/home-assistant/homeassistant/helpers/entity.py", line 349, in async_device_update
    await self.hass.async_add_executor_job(self.update)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/ehendrix/Documents/GitHub/home-assistant/homeassistant/components/media_player/directv.py", line 148, in update
    self._is_standby = self.dtv.get_standby()
AttributeError: 'NoneType' object has no attribute 'get_standby'
```
Still trying to figure this out. Looks to me that what is provided is just a MockClass and not the actual MockDirectvClass.

MartinHjelmare committed:
Remove not needed updates and assertions

MartinHjelmare committed:
Return mocked dtv instance from side_effect

MartinHjelmare committed:
Fix return correct fixture instance

MartinHjelmare committed:
Clean up assertions

MartinHjelmare committed:
Fix remaining patches

MartinHjelmare committed:
Mock time when setting up component in fixture

MartinHjelmare committed:
Patch time correctly

MartinHjelmare committed:
Attribute _last_update should return utcnow

MartinHjelmare commented:
I pushed some fixes.

ehendrix23 commented:
@MartinHjelmare 

Back. :-) Was spending some time with family over the break and was "banned" from my computer. :-)

Thank you SO much for your help in this. Been going through the changes you've made to ensure I understand as well what you did and I believe I pretty much get it. :-)
I do see that you execute async_fire_time_changed before each test. Is that to ensure that HASS does a poll for status update?

I had the asserts for hass.data[DATA_DIRECTV] in there to ensure that the information for this is added correctly as it is then used at other times to prevent duplicate entities from being created. 
Thus there is the test that makes sure that no duplicates are created based on the data in hass.data[DATA_DIRECTV] and then there are the asserts to make sure that hass.data[DATA_DIRECTV] is updated accurately. My personal preference would be that those are in there. :-)

Let me know on the asserts. Test here ran fine. Guessing I will need to do a small commit/push to remove the "Changes Requested".

Thx.

MartinHjelmare commented:
Yes, we fire a time changed event to make the scheduled polling trigger.

The state machine is the truth for the entities. Only case I've come across where we need to test hass.data when testing platforms is leftover dispatcher signals. That's not the case here I believe.

ehendrix23 commented:
As you state, the state machine is the truth of the entities. Yet within the platform we use hass.data to determine if the device (i.e. discovered) is already created or not.

Would it thus be better to include the data required to check if a duplicate exist as attributes of the entity and then to retrieve those attributes for all entities of the platform? And thus not relying on hass.data information?

MartinHjelmare commented:
If we think this is good, just remove WIP from PR title and we can merge.

ehendrix23 commented:
Thanks again for all the help. Will be using what was done here as examples for any next test I would write. :-)

MartinHjelmare commented:
If we would create a duplicate entity, there would be two states right? We can test that using the state machine. We shouldn't change the code under test. Using hass.data in the code under test is fine. Not in the tests.
